Author: Andres Salomon <dilinger@debian.org>

error[E0599]: no method named `file_as_c_str` found for reference `&Location<'_>` in the current scope
   --> ../../base/logging/logger.rs:133:39
    |
133 |         info.location().map(|loc| loc.file_as_c_str()).unwrap_or(c"<unknown file>"),
    |                                       ^^^^^^^^^^^^^ method not found in `&Location<'_>`


--- a/base/logging/rust_logger/custom_panic_hook.rs
+++ b/base/logging/rust_logger/custom_panic_hook.rs
@@ -4,6 +4,7 @@
 
 use crate::print_rust_log;
 use std::panic::PanicHookInfo;
+use std::ffi::CStr;
 
 /// Routes `panic!` and similar macros to `LOG(FATAL)`.
 pub(crate) fn init() {
@@ -14,10 +15,21 @@ pub(crate) fn init() {
     std::panic::set_hook(Box::new(panic_hook));
 }
 
+#[cfg(any(target_arch="x86", target_arch="x86_64", target_arch="loongarch64"))]
 fn panic_hook(info: &PanicHookInfo<'_>) {
     print_rust_log::print_rust_log(
         &format_args!("{info}"),
-        info.location().map(|loc| loc.file_as_c_str()).unwrap_or(c"<unknown file>"),
+        info.location().map(|loc| unsafe { CStr::from_ptr(loc.file().as_ptr() as *const i8) } ).unwrap_or(c"<unknown file>"),
+        info.location().map(|loc| loc.line()),
+        print_rust_log::LogSeverity::Fatal,
+    )
+}
+
+#[cfg(any(target_arch="powerpc64", target_arch="arm", target_arch="aarch64"))]
+fn panic_hook(info: &PanicHookInfo<'_>) {
+    print_rust_log::print_rust_log(
+        &format_args!("{info}"),
+        info.location().map(|loc| unsafe { CStr::from_ptr(loc.file().as_ptr() as *const u8) } ).unwrap_or(c"<unknown file>"),
         info.location().map(|loc| loc.line()),
         print_rust_log::LogSeverity::Fatal,
     )
--- a/base/logging/rust_logger/lib.rs
+++ b/base/logging/rust_logger/lib.rs
@@ -1,6 +1,7 @@
 // Copyright 2024 The Chromium Authors
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
+#![feature(str_as_str)]
 
 mod custom_panic_hook;
 mod log_crate_integration;
